In [1]:
import pandas as pd
import numpy as np
import math
from pprint import pprint
import dataClean as dc
import matplotlib.pyplot as plt
import seaborn as sns

## Set path as per pickle file location

In [2]:
path= 'LandResParselID.pkl'

In [3]:
noisyDF= pd.read_pickle(path)
noisyDF.head()

Parcel                                         Price Hist Street  \
0  R8394770330  {'2019': '2,595,200', '2018': '2,521,100', '20...    Not   
1  R8394770010  {'2019': '431,800', '2018': '377,300', '2017':...    Not   
2  R8394770030  {'2019': '355,900', '2018': '309,800', '2017':...    Not   
3  R8394770050  {'2019': '403,900', '2018': '352,800', '2017':...    Not   
4  R8394770070  {'2019': '0', '2018': '0', '2017': '0', '2016'...    Not   

  SideWalks Curb-Gutters Corner isCommercial Year Built Commercial Group Type  \
0                      N                   1                           OFFICE   
1                      N                   1       2014                         
2                      N                   1       2014                         
3                      N                   1       2014                         
4                      N                   1                                    

  Remodel Year  ... Pool Sq Ft Total Sq Ft Main Floor Sq Ft Upper Floor Sq Ft  \
0            0  ...                                                             
1               ...          0        1879             1444               435   
2               ...          0        1562             1562                 0   
3               ...          0        1686             1109               577   
4               ...                                                             

  Finished Lower Sq Ft Unfinished Lower Sq Ft Finished Basement Floor Sq Ft  \
0                                                                             
1                                                                             
2                                                                             
3                                                                             
4                                                                             

  Unfinished Basement Floor Sq Ft Finished Attic Sq Ft Unfinished Attic Sq Ft  
0                                                                              
1                               0                    0                         
2                               0                    0                         
3                               0                    0                         
4                                                                              

[5 rows x 37 columns]

## Missing Data Analysis

In [4]:
pprint(dc.checkNAN(noisyDF))

'You got a clean Dataframe'


In [5]:
noisyDF= noisyDF.replace('', math.nan)

In [6]:
pprint(dc.checkNAN(noisyDF))

{'Air Conditioning': '20.1784%',
 'Car Storage Sq Ft #1': '20.2750%',
 'Car Storage Sq Ft #2': '20.2750%',
 'Commercial Group Type': '95.1384%',
 'Corner': '100.0000%',
 'Curb-Gutters': '4.2536%',
 'Deck Covered': '20.2750%',
 'Deck Sq Ft': '19.5231%',
 'Finished Attic Sq Ft': '20.2750%',
 'Finished Basement Floor Sq Ft': '100.0000%',
 'Finished Lower Sq Ft': '100.0000%',
 'General Purpose Building Sq Ft': '20.2750%',
 'Ground Floor Sq Ft': '95.1384%',
 'Leasable Sq Ft': '95.1384%',
 'Main Floor Sq Ft': '20.2750%',
 'Number of Bathrooms': '19.3866%',
 'Number of Bedrooms': '20.1422%',
 'Number of FirePlaces': '20.2750%',
 'Number of Floors': '99.4975%',
 'Number of Units': '95.1384%',
 'Patio Covered': '20.2750%',
 'Patio Sq Ft': '20.2750%',
 'Pool Sq Ft': '20.2750%',
 'Porch Sq Ft': '20.2750%',
 'Remodel Year': '95.1384%',
 'SideWalks': '14.1233%',
 'Street': '3.8115%',
 'Total Sq Ft': '19.5730%',
 'Total Square Feet': '95.1384%',
 'Unfinished Attic Sq Ft': '99.6792%',
 'Unfinished Ba

**Dropping features missing 100% data**

In [7]:
noisyDF= noisyDF.drop('Unfinished Lower Sq Ft', axis=1).drop('Finished Lower Sq Ft', axis=1).drop('Finished Basement Floor Sq Ft', axis=1).drop('Corner', axis=1)

In [8]:
pprint(dc.checkNAN(noisyDF))

{'Air Conditioning': '20.1784%',
 'Car Storage Sq Ft #1': '20.2750%',
 'Car Storage Sq Ft #2': '20.2750%',
 'Commercial Group Type': '95.1384%',
 'Curb-Gutters': '4.2536%',
 'Deck Covered': '20.2750%',
 'Deck Sq Ft': '19.5231%',
 'Finished Attic Sq Ft': '20.2750%',
 'General Purpose Building Sq Ft': '20.2750%',
 'Ground Floor Sq Ft': '95.1384%',
 'Leasable Sq Ft': '95.1384%',
 'Main Floor Sq Ft': '20.2750%',
 'Number of Bathrooms': '19.3866%',
 'Number of Bedrooms': '20.1422%',
 'Number of FirePlaces': '20.2750%',
 'Number of Floors': '99.4975%',
 'Number of Units': '95.1384%',
 'Patio Covered': '20.2750%',
 'Patio Sq Ft': '20.2750%',
 'Pool Sq Ft': '20.2750%',
 'Porch Sq Ft': '20.2750%',
 'Remodel Year': '95.1384%',
 'SideWalks': '14.1233%',
 'Street': '3.8115%',
 'Total Sq Ft': '19.5730%',
 'Total Square Feet': '95.1384%',
 'Unfinished Attic Sq Ft': '99.6792%',
 'Unfinished Basement Floor Sq Ft': '20.2750%',
 'Upper Floor Sq Ft': '20.2750%',
 'Year Built': '19.2055%'}


**From the list above, I will highlight a few of the fields for which boolean features can be generated.**
   1.  'Air Conditioning': 'hasAirConditioning'
   2.  'Car Storage Sq Ft #1': 'hasCarStorage1'
   3.  'Car Storage Sq Ft #2': 'hasCarStorage2'
   4.  'Commercial Group Type': 
   5.  'Deck Covered': 'isDeckCovered'
   6.  'Deck Sq Ft': 'hasDeck'
   7.  'Finished Attic Sq Ft': 'hasFinishedAttic'
   8.  'Ground Floor Sq Ft': 'hasGroundFloor'
   9.  'Leasable Sq Ft': 'hasLeasableGround'
   10. 'Number of Bathrooms': 'hasBathroom'
   11. 'Number of Bedrooms': 'hasBedrooms'
   12. 'Numer of Fireplaces': 'hasFireplace'
   13. 'Number of Floors': 'hasMoreFloors'
   14. 'Number of Units': 'hasMoreUnits'
   15. 'Patio Covered': 'isPatioCovered'
   16. 'Patio Sq Ft': 'hasPatio'
   17. 'Pool Sq Ft': 'hasPool'
   18. 'Porch Sq Ft': 'hasPorch'
   19. 'Remodel Year': 'isRemodeled'
   20. 'SideWalks': 'hasSideWalk'
   21. 'Street': 'isStreetFacing'
   22. 'Upper Floor Sq Ft': 'hasUpperFloor'
   
## Generating New Features

In [9]:
newCols= {'Air Conditioning': 'hasAirConditioning',
          'Car Storage Sq Ft #1': 'hasCarStorage1',
          'Car Storage Sq Ft #2': 'hasCarStorage2',
          'Deck Covered': 'isDeckCovered',
          'Deck Sq Ft': 'hasDeck',
          'Finished Attic Sq Ft': 'hasFinishedAttic',
          'Ground Floor Sq Ft': 'hasGroundFloor',
          'Leasable Sq Ft': 'hasLeasableGround',
          'Number of Bathrooms': 'hasBathroom',
          'Number of Bedrooms': 'hasBedrooms',
          'Number of FirePlaces': 'hasFireplace',
          'Number of Floors': 'hasMoreFloors',
          'Number of Units': 'hasMoreUnits',
          'Patio Covered': 'isPatioCovered',
          'Patio Sq Ft': 'hasPatio',
          'Pool Sq Ft': 'hasPool',
          'Porch Sq Ft': 'hasPorch',
          'Remodel Year': 'isRemodeled',
          'SideWalks': 'hasSideWalk',
          'Street': 'isStreetFacing',
          'Upper Floor Sq Ft': 'hasUpperFloor'}

In [10]:
noisyDF= noisyDF.replace(math.nan, 'nan')

In [11]:
for col, boolCol in newCols.items():
    noisyDF[boolCol]= noisyDF[col].apply(lambda x: 'False' if x == 'nan' else 'True')

In [12]:
pprint(noisyDF.columns)

Index(['Parcel', 'Price Hist', 'Street', 'SideWalks', 'Curb-Gutters',
       'isCommercial', 'Year Built', 'Commercial Group Type', 'Remodel Year',
       'Number of Floors', 'Number of Units', 'Leasable Sq Ft',
       'Ground Floor Sq Ft', 'Total Square Feet', 'Number of Bedrooms',
       'Number of Bathrooms', 'Number of FirePlaces', 'Air Conditioning',
       'Car Storage Sq Ft #1', 'Car Storage Sq Ft #2',
       'General Purpose Building Sq Ft', 'Porch Sq Ft', 'Deck Sq Ft',
       'Deck Covered', 'Patio Sq Ft', 'Patio Covered', 'Pool Sq Ft',
       'Total Sq Ft', 'Main Floor Sq Ft', 'Upper Floor Sq Ft',
       'Unfinished Basement Floor Sq Ft', 'Finished Attic Sq Ft',
       'Unfinished Attic Sq Ft', 'hasAirConditioning', 'hasCarStorage1',
       'hasCarStorage2', 'isDeckCovered', 'hasDeck', 'hasFinishedAttic',
       'hasGroundFloor', 'hasLeasableGround', 'hasBathroom', 'hasBedrooms',
       'hasFireplace', 'hasMoreFloors', 'hasMoreUnits', 'isPatioCovered',
       'hasPatio', 'has

In [13]:
noisyDF['hasUpperFloor'].unique()

array(['False', 'True'], dtype=object)

Now that we have new features, we need to decide what to do with nan's. Let's look at features missing more than **50%** of their data

In [14]:
noisyDF= noisyDF.replace('nan',math.nan)

In [15]:
majorMissing= dc.majorMissing(noisyDF)
pprint(majorMissing)

{'Commercial Group Type': '95.1384%',
 'Ground Floor Sq Ft': '95.1384%',
 'Leasable Sq Ft': '95.1384%',
 'Number of Floors': '99.4975%',
 'Number of Units': '95.1384%',
 'Remodel Year': '95.1384%',
 'Total Square Feet': '95.1384%',
 'Unfinished Attic Sq Ft': '99.6792%'}


-----------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------
6 of the 8 features are missing same percent of data, **95.1384%**. One of the question coming to my mind is **Are they missing values for the same houses?**
For this, let's look at if they have the same *index*.

In [16]:
missingIndices= {'Commercial Group Type':list(),
                'Ground Floor Sq Ft':list(),
                'Leasable Sq Ft': list(),
                'Number of Units':list(),
                'Remodel Year':list(),
                'Total Square Feet':list()}
for key in missingIndices.keys():
    missingIndices[key]= noisyDF[noisyDF[key].isna()].index

In [17]:
missingIndices

{'Commercial Group Type': Int64Index([     1,      2,      3,      4,      5,      6,      7,      8,
                  9,     10,
             ...
             190441, 190442, 190443, 190444, 190445, 190446, 190447, 190448,
             190449, 190450],
            dtype='int64', length=181192),
 'Ground Floor Sq Ft': Int64Index([     1,      2,      3,      4,      5,      6,      7,      8,
                  9,     10,
             ...
             190441, 190442, 190443, 190444, 190445, 190446, 190447, 190448,
             190449, 190450],
            dtype='int64', length=181192),
 'Leasable Sq Ft': Int64Index([     1,      2,      3,      4,      5,      6,      7,      8,
                  9,     10,
             ...
             190441, 190442, 190443, 190444, 190445, 190446, 190447, 190448,
             190449, 190450],
            dtype='int64', length=181192),
 'Number of Units': Int64Index([     1,      2,      3,      4,      5,      6,      7,      8,
                  9,

In [18]:
def checkIfEqual(lst1, lst2):
    return sorted(lst1) == sorted(lst2)

In [19]:
i=0
prev= ''
for key,col in missingIndices.items():
    if i == 0:
        prev= key
        prevCol= col
        i+= 1
    else:
        print(key,',', prev,'-->',checkIfEqual(col,prevCol))
        prev= key
        prevCol= col
        i+= 1

Ground Floor Sq Ft , Commercial Group Type --> True
Leasable Sq Ft , Ground Floor Sq Ft --> True
Number of Units , Leasable Sq Ft --> True
Remodel Year , Number of Units --> True
Total Square Feet , Remodel Year --> True


So these features are missing data for same houses.
***So if a house doesn't have Ground Floor, Leasable Sq Ft, Number of Units,Remodel Year, Total Square Feet, what does it imply?***

--------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------
Looking at majorMissing again, these features are missing almost all of their data so dropping these features.

In [20]:
for key in majorMissing.keys():
    noisyDF= noisyDF.drop(key, axis= 1)

In [21]:
len(noisyDF.columns)

46

In [22]:
missing= dc.checkNAN(noisyDF)
pprint(missing)

{'Air Conditioning': '20.1784%',
 'Car Storage Sq Ft #1': '20.2750%',
 'Car Storage Sq Ft #2': '20.2750%',
 'Curb-Gutters': '4.2536%',
 'Deck Covered': '20.2750%',
 'Deck Sq Ft': '19.5231%',
 'Finished Attic Sq Ft': '20.2750%',
 'General Purpose Building Sq Ft': '20.2750%',
 'Main Floor Sq Ft': '20.2750%',
 'Number of Bathrooms': '19.3866%',
 'Number of Bedrooms': '20.1422%',
 'Number of FirePlaces': '20.2750%',
 'Patio Covered': '20.2750%',
 'Patio Sq Ft': '20.2750%',
 'Pool Sq Ft': '20.2750%',
 'Porch Sq Ft': '20.2750%',
 'SideWalks': '14.1233%',
 'Street': '3.8115%',
 'Total Sq Ft': '19.5730%',
 'Unfinished Basement Floor Sq Ft': '20.2750%',
 'Upper Floor Sq Ft': '20.2750%',
 'Year Built': '19.2055%'}


In [23]:
noisyDF[missing.keys()].describe()

Street SideWalks Curb-Gutters Year Built Number of Bedrooms  \
count             183192    163553       182350     153874             152090   
unique                 4         2            2        141                 14   
top     Asphalt/Concrete         Y            Y       2005                  3   
freq              140633    112541       119985       6650              79096   

       Number of Bathrooms Number of FirePlaces Air Conditioning  \
count               153529               151837           152021   
unique                  46                    8                2   
top                    2.0                    1                Y   
freq                 57585                94963           139314   

       Car Storage Sq Ft #1 Car Storage Sq Ft #2  ... Deck Sq Ft Deck Covered  \
count                151837               151837  ...     153269       151837   
unique                 2052                 1104  ...       1001            2   
top                       0                    0  ...          0            N   
freq                   9525               144894  ...     126678       144372   

       Patio Sq Ft Patio Covered Pool Sq Ft Total Sq Ft Main Floor Sq Ft  \
count       151837        151837     151837      153174           151837   
unique        1346             2        447        5659             3997   
top              0             N          0           0                0   
freq         20976         82773     147722        1584             1587   

       Upper Floor Sq Ft Unfinished Basement Floor Sq Ft Finished Attic Sq Ft  
count             151837                          151837               151837  
unique              2389                            1353                 1032  
top                    0                               0                    0  
freq              101392                          145414               140426  

[4 rows x 22 columns]

Some of these features nan's value correspond to N. Such as **SideWalks, Curb Gutters, AirConditioning, Deck Covered, Patio Covered.**
We already have boolean fields for **AirConditioning, Deck Covered, Patio Covered.** So reducing redundancy in our data and dropping these features.

In [24]:
noisyDF= noisyDF.drop('Air Conditioning', axis=1).drop('Deck Covered',axis=1).drop('Patio Covered', axis=1)

In [25]:
missing= dc.checkNAN(noisyDF)
pprint(missing)

{'Car Storage Sq Ft #1': '20.2750%',
 'Car Storage Sq Ft #2': '20.2750%',
 'Curb-Gutters': '4.2536%',
 'Deck Sq Ft': '19.5231%',
 'Finished Attic Sq Ft': '20.2750%',
 'General Purpose Building Sq Ft': '20.2750%',
 'Main Floor Sq Ft': '20.2750%',
 'Number of Bathrooms': '19.3866%',
 'Number of Bedrooms': '20.1422%',
 'Number of FirePlaces': '20.2750%',
 'Patio Sq Ft': '20.2750%',
 'Pool Sq Ft': '20.2750%',
 'Porch Sq Ft': '20.2750%',
 'SideWalks': '14.1233%',
 'Street': '3.8115%',
 'Total Sq Ft': '19.5730%',
 'Unfinished Basement Floor Sq Ft': '20.2750%',
 'Upper Floor Sq Ft': '20.2750%',
 'Year Built': '19.2055%'}


Differentiating between discrete and continous data. Looking at the number of unique values for each feature.

In [26]:
categs= dc.getCateg(noisyDF, missing.keys())
for k, v in categs.items():
    print(k, ':',f'{(len(v)/len(noisyDF[k]))*100:.3g}','%')

Street : 0.00263 %
SideWalks : 0.00158 %
Curb-Gutters : 0.00158 %
Year Built : 0.0746 %
Number of Bedrooms : 0.00788 %
Number of Bathrooms : 0.0247 %
Number of FirePlaces : 0.00473 %
Car Storage Sq Ft #1 : 1.08 %
Car Storage Sq Ft #2 : 0.58 %
General Purpose Building Sq Ft : 0.472 %
Porch Sq Ft : 0.41 %
Deck Sq Ft : 0.526 %
Patio Sq Ft : 0.707 %
Pool Sq Ft : 0.235 %
Total Sq Ft : 2.97 %
Main Floor Sq Ft : 2.1 %
Upper Floor Sq Ft : 1.25 %
Unfinished Basement Floor Sq Ft : 0.711 %
Finished Attic Sq Ft : 0.542 %


Looking at the proportion of unique values in the whole dataset, I am inclined to treat these features as categorical. ***But should we really?***

Treating them as **categorical.** Using *fixCateg* function from *dataClean.py*
The fixCateg function replaces the nan taking into consideration the proportion of classes of the features. Weights are assigned to each class. The dominant class will have a lower weight and vice versa.

I thought of different approaches. Replace nan's with the mode of the class, or replace nan's with median of the class (if discrete). 

In [27]:
lessNoisyDF= dc.fixCateg(noisyDF, missing.keys())

In [28]:
dc.checkNAN(lessNoisyDF)

'You got a clean Dataframe'

In [29]:
lessNoisyDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190451 entries, 0 to 190450
Data columns (total 43 columns):
Parcel                             190451 non-null object
Price Hist                         190451 non-null object
Street                             190451 non-null object
SideWalks                          190451 non-null object
Curb-Gutters                       190451 non-null object
isCommercial                       190451 non-null object
Year Built                         190451 non-null object
Number of Bedrooms                 190451 non-null object
Number of Bathrooms                190451 non-null object
Number of FirePlaces               190451 non-null object
Car Storage Sq Ft #1               190451 non-null object
Car Storage Sq Ft #2               190451 non-null object
General Purpose Building Sq Ft     190451 non-null object
Porch Sq Ft                        190451 non-null object
Deck Sq Ft                         190451 non-null object
Patio Sq Ft        

Now we have a clean dataframe. Let's set our target variable.

In [30]:
lessNoisyDF['Price Hist'].head()

0    {'2019': '2,595,200', '2018': '2,521,100', '20...
1    {'2019': '431,800', '2018': '377,300', '2017':...
2    {'2019': '355,900', '2018': '309,800', '2017':...
3    {'2019': '403,900', '2018': '352,800', '2017':...
4    {'2019': '0', '2018': '0', '2017': '0', '2016'...
Name: Price Hist, dtype: object

In [31]:
lessNoisyDF['Price']= lessNoisyDF['Price Hist'].apply(lambda x: x.get('2019'))

In [32]:
lessNoisyDF.columns

Index(['Parcel', 'Price Hist', 'Street', 'SideWalks', 'Curb-Gutters',
       'isCommercial', 'Year Built', 'Number of Bedrooms',
       'Number of Bathrooms', 'Number of FirePlaces', 'Car Storage Sq Ft #1',
       'Car Storage Sq Ft #2', 'General Purpose Building Sq Ft', 'Porch Sq Ft',
       'Deck Sq Ft', 'Patio Sq Ft', 'Pool Sq Ft', 'Total Sq Ft',
       'Main Floor Sq Ft', 'Upper Floor Sq Ft',
       'Unfinished Basement Floor Sq Ft', 'Finished Attic Sq Ft',
       'hasAirConditioning', 'hasCarStorage1', 'hasCarStorage2',
       'isDeckCovered', 'hasDeck', 'hasFinishedAttic', 'hasGroundFloor',
       'hasLeasableGround', 'hasBathroom', 'hasBedrooms', 'hasFireplace',
       'hasMoreFloors', 'hasMoreUnits', 'isPatioCovered', 'hasPatio',
       'hasPool', 'hasPorch', 'isRemodeled', 'hasSideWalk', 'isStreetFacing',
       'hasUpperFloor', 'Price'],
      dtype='object')

The **PriceHist** column has a dictionary which has the past prices for the house. So we need to *explode* this dictionary and have the keys as columns, eg. PriceIn2018, PriceIn2017....

In [33]:
yearLst= ['2018','2017','2016','2015','2014','2013','2012','2011','2010','2009','2008']

In [34]:
for year in yearLst:
    colName= 'PriceIn'+year
    lessNoisyDF[colName]= lessNoisyDF['Price Hist'].apply(lambda x: x.get(year))

In [35]:
lessNoisyDF.columns

Index(['Parcel', 'Price Hist', 'Street', 'SideWalks', 'Curb-Gutters',
       'isCommercial', 'Year Built', 'Number of Bedrooms',
       'Number of Bathrooms', 'Number of FirePlaces', 'Car Storage Sq Ft #1',
       'Car Storage Sq Ft #2', 'General Purpose Building Sq Ft', 'Porch Sq Ft',
       'Deck Sq Ft', 'Patio Sq Ft', 'Pool Sq Ft', 'Total Sq Ft',
       'Main Floor Sq Ft', 'Upper Floor Sq Ft',
       'Unfinished Basement Floor Sq Ft', 'Finished Attic Sq Ft',
       'hasAirConditioning', 'hasCarStorage1', 'hasCarStorage2',
       'isDeckCovered', 'hasDeck', 'hasFinishedAttic', 'hasGroundFloor',
       'hasLeasableGround', 'hasBathroom', 'hasBedrooms', 'hasFireplace',
       'hasMoreFloors', 'hasMoreUnits', 'isPatioCovered', 'hasPatio',
       'hasPool', 'hasPorch', 'isRemodeled', 'hasSideWalk', 'isStreetFacing',
       'hasUpperFloor', 'Price', 'PriceIn2018', 'PriceIn2017', 'PriceIn2016',
       'PriceIn2015', 'PriceIn2014', 'PriceIn2013', 'PriceIn2012',
       'PriceIn2011', 'PriceI

In [36]:
lessNoisyDF.shape

(190451, 55)

## Write Dataset to file

In [37]:
lessNoisyDF.to_pickle('CleanSet.pkl')